### PR TITLE
[util, sw] Change format of generated register definitions

### PR DESCRIPTION
**Pull request by silvestrst**
regtool.py generates register definitions with assumed knowledge of
base_address, however the DIF specification (#1043) prohibits any
knowledge of adresses belonging to hardware blocks. Instead this
information (such as base_address) has to be passed into the dif by
the consumer.

This change modifies the gen_cheader.py to only produce relative
module addresses (register offsets from the base), and defines the
absolute register addresses in the driver (base + relative).

The approach taken - keep the format <device>_<register>(id) to avoid
changing all the places where these macros are called.

With introduction of DIFs all these definitions will hopefully become
redundant.

Easy alternative would be to keep the generated format, define "fake"
<device><n>_BASE_ADDR and then calculate the register address inside
the DIF by adding passed in "real" base address with the register
offset. This way the current implementation can remain untouched, and
when DIFs are completed, gen_cheader.py can be changed.

This change fixes issue #1083

Signed-off-by: Silvestrs Timofejevs <silvestrst@lowrisc.org>

### CODE REVIEW

**Comment by silvestrst** [comment]
I think the alternative is probably better? Or maybe some other ideas?

**Comment by tjaychen** [comment]
sorry i'm a bit late to the party.  It does feel like the alternative you described is better, and also saves a lot of redundant definition in the higher level files. 

There was another related question posed in the DIF specification, which is how we handle HW blocks that are paramterized slightly differently.  I assume it would be similar? Another parameter passed in to differentiate behavior?

File: sw/device/examples/hello_usbdev/hello_usbdev.c [hunk 1]
```diff
@@ -8,10 +8,14 @@
 #include "sw/device/lib/usb_controlep.h"
 #include "sw/device/lib/usb_simpleserial.h"
 #include "sw/device/lib/usbdev.h"
+#include "usbdev_regs.h"  // Generated.
 
 // These just for the '/' printout
 #define USBDEV_BASE_ADDR 0x40020000
-#include "usbdev_regs.h"  // Generated.
+
+#define USBDEV_REG_DEF(id, rname) (REG_DEF(USBDEV, id, rname))
```
Comment (mcy) [inline sw/device/examples/hello_usbdev/hello_usbdev.c]: I'd get rid of these `_REG_DEF` macros. They don't really save much on typing and are extra, unnecessary indirection.

File: sw/device/lib/common.h [hunk 2]
```diff
@@ -25,6 +25,8 @@ static const unsigned long UART_BAUD_RATE = 230400;
 #define FLASH_PAGES_PER_BANK 256
 #define FLASH_BANK_SZ FLASH_PAGES_PER_BANK *FLASH_PAGE_SZ
 
+#define REG_DEF(comp, id, rname) (comp##id##_BASE_ADDR + comp##_##rname##_REG)
```
Comment (mcy) [inline sw/device/lib/common.h]: This macro wants documentation, and something like "this is for compatibility only do not use in new code".

**Comment by silvestrst** [comment]
> sorry i'm a bit late to the party. It does feel like the alternative you described is better, and also saves a lot of redundant definition in the higher level files.
> 
> There was another related question posed in the DIF specification, which is how we handle HW blocks that are paramterized slightly differently. I assume it would be similar? Another parameter passed in to differentiate behavior?

**Comment by silvestrst** [comment]
> sorry i'm a bit late to the party. It does feel like the alternative you described is better, and also saves a lot of redundant definition in the higher level files.
> 
> There was another related question posed in the DIF specification, which is how we handle HW blocks that are paramterized slightly differently. I assume it would be similar? Another parameter passed in to differentiate behavior?

Do you mean same types of HW block, like (uart0, uart1), or just HW blocks that don't fit in with the generic way of how register definitions are laid out for the other HW blocks?

File: sw/device/examples/hello_usbdev/hello_usbdev.c (hunk 1 above)
Comment (silvestrst) [reply sw/device/examples/hello_usbdev/hello_usbdev.c]: I agree, and will update the change if we decide to go with this PR.

File: sw/device/lib/common.h (hunk 2 above)
Comment (silvestrst) [reply sw/device/lib/common.h]: I agree, and will update the change if we decide to go with this PR.

**Comment by silvestrst** [comment]
Now when there is consensus that *_regs.h should only be accessible by the corresponding peripheral DIF, I think it makes sense to go for this change. As a part of this change, we can also remove all dependencies on the *_regs.h from the consumers. What do you think @gkelly @mcy @lenary @tjaychen ? If you are happy with this I can do it on Monday (13/12/2019).

**Comment by mcy** [comment]
I think this is fine. I still need to work out how these headers are going to be generated by Meson, because the status quo is kind-of unacceptable (they're just generated at the meson root rather than in a reasonable directory).

**Comment by silvestrst** [comment]
Actually thinking about it again - in order to avoid "throw away" work, I would rather like to wait until #1187 lands, and use reg32_* functions. This way when the DIFs for the modules are implemented, it will be a matter of simple substitution to get the address from dif_device structure instead.

**Comment by silvestrst** [comment]
Closing this PR, in favour of #1118

**Pull request closed without merging**